File: examples/typescript/property-tests.ts
// Property-based testing with node:test and manual random generation.
import { describe, it } from "node:test";
import assert from "node:assert/strict";

// --- Functions under test ---

function applyDiscount(cents: number, pct: number): number {
  const clampedPct = Math.max(0, Math.min(100, Math.floor(pct)));
  return Math.floor(cents * (100 - clampedPct) / 100);
}

function addMoney(a: number, b: number): number {
  return a + b;
}

// --- Simple RNG for reproducible tests ---

function randomInts(seed: number, count: number, max: number): number[] {
  const values: number[] = [];
  let state = seed;
  for (let i = 0; i < count; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    values.push(state % max);
  }
  return values;
}

const ITERATIONS = 200;

describe("property: applyDiscount", () => {
  it("result is bounded between 0 and original", () => {
    const cents = randomInts(42, ITERATIONS, 1_000_000);
    const pcts = randomInts(43, ITERATIONS, 101);
    for (let i = 0; i < ITERATIONS; i++) {
      const result = applyDiscount(cents[i], pcts[i]);
      assert.ok(result >= 0, `result ${result} should be >= 0`);
      assert.ok(result <= cents[i], `result ${result} should be <= ${cents[i]}`);
    }
  });

  it("0% discount is identity", () => {
    for (const c of randomInts(44, ITERATIONS, 1_000_000)) {
      assert.equal(applyDiscount(c, 0), c);
    }
  });

  it("100% discount is zero", () => {
    for (const c of randomInts(45, ITERATIONS, 1_000_000)) {
      assert.equal(applyDiscount(c, 100), 0);
    }
  });

  it("monotonic: higher discount → lower result", () => {
    const cents = randomInts(46, ITERATIONS, 1_000_000);
    const lo = randomInts(47, ITERATIONS, 100);
    const hi = randomInts(48, ITERATIONS, 100);
    for (let i = 0; i < ITERATIONS; i++) {
      const [a, b] = lo[i] <= hi[i] ? [lo[i], hi[i]] : [hi[i], lo[i]];
      assert.ok(applyDiscount(cents[i], a) >= applyDiscount(cents[i], b));
    }
  });
});

describe("property: addMoney", () => {
  it("commutative", () => {
    const a = randomInts(49, ITERATIONS, 100_000);
    const b = randomInts(50, ITERATIONS, 100_000);
    for (let i = 0; i < ITERATIONS; i++) {
      assert.equal(addMoney(a[i], b[i]), addMoney(b[i], a[i]));
    }
  });
});

File: examples/rust/property_tests.rs
//! Property-based testing with stdlib (manual random generation).
//!
//! For production Rust, use `proptest` crate for property-based testing.
//! This shows the pattern with stdlib only.
#![allow(dead_code)]

fn apply_discount(cents: u64, pct: u8) -> u64 {
    let pct = pct.min(100);
    cents * (100 - pct as u64) / 100
}

fn clamp(value: i64, min: i64, max: i64) -> i64 {
    if value < min {
        min
    } else if value > max {
        max
    } else {
        value
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn simple_rng(seed: u64, count: usize) -> Vec<u64> {
        let mut values = Vec::with_capacity(count);
        let mut state = seed;
        for _ in 0..count {
            state = state.wrapping_mul(6364136223846793005).wrapping_add(1);
            values.push(state >> 33);
        }
        values
    }

    const ITERATIONS: usize = 500;

    #[test]
    fn discount_bounded() {
        for v in simple_rng(42, ITERATIONS * 2).chunks(2) {
            let cents = v[0] % 1_000_000;
            let pct = (v[1] % 101) as u8;
            let result = apply_discount(cents, pct);
            assert!(result <= cents, "discount should not increase price");
        }
    }

    #[test]
    fn discount_zero_is_identity() {
        for cents in simple_rng(123, ITERATIONS) {
            let cents = cents % 1_000_000;
            assert_eq!(apply_discount(cents, 0), cents);
        }
    }

    #[test]
    fn discount_hundred_is_zero() {
        for cents in simple_rng(456, ITERATIONS) {
            let cents = cents % 1_000_000;
            assert_eq!(apply_discount(cents, 100), 0);
        }
    }

    #[test]
    fn discount_monotonic() {
        for v in simple_rng(789, ITERATIONS * 3).chunks(3) {
            let cents = v[0] % 1_000_000;
            let (lo, hi) = if v[1] % 101 <= v[2] % 101 {
                ((v[1] % 101) as u8, (v[2] % 101) as u8)
            } else {
                ((v[2] % 101) as u8, (v[1] % 101) as u8)
            };
            assert!(apply_discount(cents, lo) >= apply_discount(cents, hi));
        }
    }

    #[test]
    fn clamp_idempotent() {
        for v in simple_rng(321, ITERATIONS * 3).chunks(3) {
            let (value, min, max) = (v[0] as i64 % 1000, 0i64, 100i64);
            let once = clamp(value, min, max);
            let twice = clamp(once, min, max);
            assert_eq!(once, twice, "clamp should be idempotent");
        }
    }
}
